### PR TITLE
OEP-18 Compliancy, compatibility tests on Ironwood

### PR DESCRIPTION
**Pull request by spokerman12**
# OEP-18 Compliancy, compatibility tests on Ironwood

## Background:
From [edx](https://open-edx-proposals.readthedocs.io/en/latest/oep-0018-bp-python-dependencies.html):
> OEP-18 is a recommended standard for declaring dependencies on other Python packages which resolves most of these issues and will let us make all the Open edX Python packages consistent with each other (and many other open source Python projects) for ease of understanding and maintenance.

This PR focuses on making the necessary changes for Flow Control to comply with OEP-18.

Additionally, we ensured (live testing, tox tests) that this XBlock works correctly with [Ironwood](https://github.com/edx/edx-platform/tree/open-release/ironwood.master).

## Context:

OEP-18 specifies 5 main points of action:

1. Identify the different contexts in which dependencies will need to be installed.
2. For each of these contexts, declare the direct dependencies that will be needed. Use the least restrictive constraints which should allow pip to install a working set of dependencies. We generally no longer want or need to manually pin specific versions; make upgrade will do this for us.
3. Run ``make upgrade`` to generate from the high-level dependencies declarations a separate set of requirements files that specify the exact package versions that are known to work for a Python virtualenv created for that context.
4. Automate updates of the detailed requirements files for each context.
5. When making deliberate changes to the repository’s dependencies, only edit the files given to *.in files in the requirements directory, never the generated *.txt files. Both types of files must be committed when changed.

Following these directions, these changes ensued:

- We studied each code dependency in this repository, which allowed us to create MANIFEST, base, test, quality, and dev ``.in`` files. 
- Changed git dependencies so instead they are downloaded from PyPI.
- With ``constraints.txt``, strapped dependency versions to match those of ``open-release/ironwood.master``.
- We found out that some dependencies had no apparent use (such as ``pep8``), so these were left as comments for the prospect of implementing, for example, quality and coverage tests.
- After cleaning up the dependencies, we created a Makefile with ``help``, ``install``, ``requirements``, ``upgrade``.
- Following OEP-18, we hooked up ``pip-compile`` with the Makefile to keep the .txt requirement files updated.

## Aditional changes

- Cleaned up README.rst
- Deleted ``circle.yml`` file, as it's obsolete for CircleCI 2.1

## Next steps

- Upgrade this XBlock so it works with Juniper and we can soften the dependency constraints.

## Reviewers

- @felipemontoya 
- @morenol

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
 ==================================
 
-The Flow Control Xblock provides a way to display the content of a unit or to redirect the user elsewhere based on compliance with a condition that evaluates the submission or the score of a problem or a set of problems.
+The Flow Control XBlock provides a way to display the content of a unit or to redirect the user elsewhere based on compliance with a condition that evaluates the submission or the score of a problem or a set of problems.
 
 Installation
 ------------
 
-Install the requirements into the Python virtual environment of your `edx-platform` installation by running the following command from the root folder:
+Inside both LMS and Studio shells, using `make lms-shell` and `make studio-shell` in your devstack directory:
```
Comment (morenol) [inline README.rst:10]: I think that these changes should be in another PR, if we are going to add this, then the title should have to change to something more related to installation on the devstack.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	pip install -r requirements/dev.txt
+
+test: ## test using tox
+	pip install tox
+	tox	
+
+upgrade: export CUSTOM_COMPILE_COMMAND=make upgrade
+upgrade: ## update the pip requirements files to use the latest releases satisfying our constraints
+	pip install -q pip-tools
+	pip-compile --rebuild --upgrade -o requirements/dev.txt requirements/dev.in
```
Comment (morenol) [inline Makefile]: We need a line for base.in/base.txt and for pip-tools.in/pip-tools.txt

File: requirements/dev.in [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Additional requirements for development of this application
+-c constraints.txt
+
+-r base.in                # Dependencies for edx-completion itself
```
Comment (morenol) [inline requirements/dev.in]: This should be base.txt in order to use the same dependencies generated by base.in in order to prevent the possiblity of different release if there is a new release just in the moment that this file is being compiled.

In general, I think that we can use just test.txt and quality.txt

File: requirements/dev.in [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+-r quality.in             # Dependencies for code quality checks
+-r test.in                # Dependencies for test case execution
+
+pip-tools                 # Requirements file management
```
Comment (morenol) [inline requirements/dev.in]: This should be in a file called pip-tools.in/pip-tools.txt

File: requirements/dev.in [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+tox                       # virtualenv management for tests
```
Comment (morenol) [inline requirements/dev.in]: Please an empty line at the end of this file and also please add this dependency to a file called tox.in/tox.txt

File: tox.ini [hunk 6]
```diff
@@ -3,7 +3,7 @@
 envlist = py27
 
 [testenv]
-deps = -rflow_control/requirements/test.txt
+deps = -r{toxinidir}/requirements/test.in
```
Comment (morenol) [inline tox.ini]: Please use the compiled file text.txt

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+requirements: ## fetch development requirements
+	pip install -r requirements/dev.txt
+
+test: ## test using tox
+	pip install tox
+	tox	
+
+upgrade: export CUSTOM_COMPILE_COMMAND=make upgrade
+upgrade: ## update the pip requirements files to use the latest releases satisfying our constraints
+	pip install -q pip-tools
```
Comment (morenol) [inline Makefile]: Please install this using the requirements/pip-tools.txt file generated by `make upgrade`

File: requirements/pip-tools.in [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+#
+# If you open a pull request that adds a new dependency, you should:
+#   * verify that the dependency has a license compatible with AGPLv3
+#   * confirm that it has no system requirements beyond what we already install
+#   * run "make upgrade" to update the detailed requirements files
+
+-c constraints.txt
+
+pip-tools                 # Contains pip-compile, used to generate pip requirements files
+six                       # Pinned because pip-tools needs it, but later transifex-client says ==1.11.0
```
Comment (morenol) [inline requirements/pip-tools.in]: I think that we can remove this dependency from here

File: requirements/constraints.txt [hunk 9]
```diff
@@ -0,0 +1,28 @@
+# Version constraints for pip installation.
+#
+# This file doesn't install any packages. It specifies version constraints
+# that will be applied if a package is needed.
+#
+# When pinning something here, please provide an explanation of why.  Ideally,
+# link to other information that will help people in the future to remove the
+# pin when possible.  Writing an issue against the offending project and
+# linking to it here is good.
+
```
Comment (morenol) [inline requirements/constraints.txt:10]: Sort the dependencies alphabetically

File: requirements/quality.in [hunk 10]
```diff
@@ -0,0 +1,6 @@
+# Code quality testing dependencies
+# These tests should be added soon.
+
```
Comment (morenol) [inline requirements/quality.in:3]: Add -c contraints.txt

File: requirements/test.in [hunk 11]
```diff
@@ -0,0 +1,13 @@
+# Requirements for testing Flow-Control
+
```
Comment (morenol) [inline requirements/test.in:2]: add -c constraints.txt

File: README.rst (hunk 1 above)
Comment (spokerman12) [reply README.rst:10]: Changed to:

Installing on Open edX Devstack
------------

File: Makefile (hunk 2 above)
Comment (spokerman12) [reply Makefile]: Line added. Removed line for ``dev.in``

File: requirements/dev.in (hunk 3 above)
Comment (spokerman12) [reply requirements/dev.in]: ``dev.in`` roles were given to ``base.in``.
deleted ``dev.in``.

File: requirements/dev.in (hunk 4 above)
Comment (spokerman12) [reply requirements/dev.in]: pip-tools.in file created. pip-tools.txt will come from pip-compile.

File: requirements/dev.in (hunk 5 above)
Comment (spokerman12) [reply requirements/dev.in]: Added tox.in file. Added to ``make upgrade`` list of commands.

File: tox.ini (hunk 6 above)
Comment (spokerman12) [reply tox.ini]: Fixed

File: Makefile (hunk 7 above)
Comment (spokerman12) [reply Makefile]: Edited accordingly.

File: requirements/pip-tools.in (hunk 8 above)
Comment (spokerman12) [reply requirements/pip-tools.in]: Removed. Now pip-tools is in pip-tools.in

File: requirements/constraints.txt (hunk 9 above)
Comment (spokerman12) [reply requirements/constraints.txt:10]: Sorted

File: requirements/quality.in (hunk 10 above)
Comment (spokerman12) [reply requirements/quality.in:3]: Added

File: requirements/test.in (hunk 11 above)
Comment (spokerman12) [reply requirements/test.in:2]: Added

File: requirements/base.in [hunk 12]
```diff
@@ -0,0 +1,7 @@
+# Base requirements for using Flow-Control
+
+-c constraints.txt
+
+xblock                              # Base courseware component architecture
+xblock-utils                        # Provides utilities used by the Flow-Control XBlock
+edx-opaque-keys                     # Allows fetch scores from Courseware's ScoresClient
```
Comment (morenol) [inline requirements/base.in]: Let's avoid the error that we had in eox-tagging, so change this with `edx-opaque-keys[django]` or move this from here to test.in

File: requirements/constraints.txt [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+ddt==1.2.0
+
+# Score fetching compatible with open-release/ironwood.master
+edx-opaque-keys==2.1.0
+
+# Testing mock tool compatible with open-release/ironwood.master
+mock==1.0.1
+
+# Nose testing for Django, although migration to Pytest should happen soon.
+nosedjango==1.1.0
```
Comment (morenol) [inline requirements/constraints.txt]: Is this constraint needed?

File: requirements/constraints.txt [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+# that will be applied if a package is needed.
+#
+# When pinning something here, please provide an explanation of why.  Ideally,
+# link to other information that will help people in the future to remove the
+# pin when possible.  Writing an issue against the offending project and
+# linking to it here is good.
+
+
+# Test case helper compatible with open-release/ironwood.master
+ddt==1.2.0
```
Comment (morenol) [inline requirements/constraints.txt]: Is this constraint needed? I think that it is only used in the tests we could try to use a newer version

File: requirements/tox.txt [hunk 15]
```diff
@@ -0,0 +1,26 @@
+#
```
Comment (morenol) [inline requirements/tox.txt:1]: Change the circleci/config.yml to use install this file instead of `pip install tox`

File: Makefile [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+	@perl -nle'print $& if m{^[a-zA-Z_-]+:.*?## .*$$}' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m  %-25s\033[0m %s\n", $$1, $$2}'
+
+install: ## install the flow-control xblock
+	pip install .
+
+requirements: ## fetch development requirements
+	pip install -r requirements/base.txt
+
+test: ## test using tox
+	pip install tox
```
Comment (morenol) [inline Makefile]: Change here to use requirements/tox.txt

Comment (spokerman12) [reply Makefile]: Changed accordingly.

File: requirements/base.in (hunk 12 above)
Comment (spokerman12) [reply requirements/base.in]: Changed to edx-opaque-keys[django]. 
Added constraint <2.0.0

File: requirements/constraints.txt (hunk 14 above)
Comment (spokerman12) [reply requirements/constraints.txt]: Agreed. Constraint removed.

File: requirements/constraints.txt (hunk 13 above)
Comment (spokerman12) [reply requirements/constraints.txt]: Agreed. It's not necessary. Removed.

File: requirements/tox.txt (hunk 15 above)
Comment (spokerman12) [reply requirements/tox.txt:1]: Changed exactly like so.

File: requirements/constraints.txt [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+# This file doesn't install any packages. It specifies version constraints
+# that will be applied if a package is needed.
+#
+# When pinning something here, please provide an explanation of why.  Ideally,
+# link to other information that will help people in the future to remove the
+# pin when possible.  Writing an issue against the offending project and
+# linking to it here is good.
+
+
+# Score fetching compatible with open-release/ironwood.master
```
Comment (morenol) [inline requirements/constraints.txt]: Add in the comment that this is the last version compatible with django1.11

File: requirements/constraints.txt [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+edx-opaque-keys<2.1.0
+
+# Testing mock tool compatible with open-release/ironwood.master
+mock==1.0.1
```
Comment (morenol) [inline requirements/constraints.txt:16]: can we remove this constraint?

File: Makefile [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+
+requirements: ## fetch development requirements
+	pip install -r requirements/base.txt
+
+test: ## test using tox
+	pip install -r requirements/tox.txt
+	tox	
+
+upgrade: export CUSTOM_COMPILE_COMMAND=make upgrade
+upgrade: ## update the pip requirements files to use the latest releases satisfying our constraints
```
Comment (morenol) [inline Makefile:18]: This is needed before all of these commandds are run
`pip install -r requirements/pip-tools.txt`

Comment (spokerman12) [reply Makefile:18]: Line added

File: requirements/constraints.txt (hunk 17 above)
Comment (spokerman12) [reply requirements/constraints.txt]: Added like so.

**Review by morenol — APPROVED** [review]
